### PR TITLE
gnrc_ipv6_nib: 6ln: send neighbor solicitation if node is unreachable

### LINKED ISSUE #14735: Global Address remains in state TNT[3] forever
#### Description
The issue described in #13838 is still there, but the original issue also uncovered a different bug that is now fixed.
So I'm creating a new issue to reduce the noise.

#### Steps to reproduce the issue

Have a node sleep for a certain period of time.
After a couple of cycles the public address of the node will not become valid anymore.
This is remedied by a restart of the border router.

<details><summary>main.c</summary>

```C
#include <stdio.h>

#include "periph/rtc.h"
#include "periph/pm.h"
#include "msg.h"
#include "net/gnrc/netif.h"

#define NETIF_NUM_MAX               (2)
#define BORDER_ROUTER_TIMEOUT_US    (1000 * US_PER_MS)

/* ifconfig */
int _gnrc_netif_config(int argc, char **argv);

static void _rtc_alarm(void* arg)
{
    (void) arg;

    pm_reboot();
    return;
}

static void _gnrc_netapi_set_all(netopt_state_t state)
{
    gnrc_netif_t* netif = NULL;
    while ((netif = gnrc_netif_iter(netif))) {
        /* retry if busy */
        while (gnrc_netapi_set(netif->pid, NETOPT_STATE, 0,
                               &state, sizeof(state)) == -EBUSY) {}
    }
}

static void pm_hibernate_for(unsigned sec)
{
    struct tm now;

    /* disable all network interfaces */
    _gnrc_netapi_set_all(NETOPT_STATE_SLEEP);

    rtc_get_time(&now);
    now.tm_sec += sec;
    rtc_set_alarm(&now, _rtc_alarm, NULL);
    pm_set(0);
}

int main(void)
{
    msg_t _main_msg_queue[4];
    msg_bus_entry_t subs[NETIF_NUM_MAX];
    unsigned count;
    gnrc_netif_t *netif;

    msg_init_queue(_main_msg_queue, ARRAY_SIZE(_main_msg_queue));

    /* enable network interfaces */
    gnrc_netif_init_devs();

    /* subscribe to all interfaces */
    count = 0;
    netif = NULL;
    while ((netif = gnrc_netif_iter(netif)) && count < NETIF_NUM_MAX) {
        msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);

        msg_bus_attach(bus, &subs[count]);
        msg_bus_subscribe(&subs[count], GNRC_IPV6_EVENT_ADDR_VALID);

        ++count;
    }

    /* wait for global address */
    msg_t m;
    do {
        if (xtimer_msg_receive_timeout(&m, BORDER_ROUTER_TIMEOUT_US) < 0) {
            puts(">> border router timeout <<");
            _gnrc_netif_config(0, NULL);
            break;
        }
    } while (ipv6_addr_is_link_local(m.content.ptr));

    /* unsubscribe */
    count = 0;
    netif = NULL;
    while ((netif = gnrc_netif_iter(netif)) && count < NETIF_NUM_MAX) {
        msg_bus_t *bus = gnrc_netif_get_bus(netif, GNRC_NETIF_BUS_IPV6);
        msg_bus_detach(bus, &subs[count++]);
    }

    pm_hibernate_for(5);
    puts(">> hibernate failed? <<");

    return 0;
}
```
</details>

<details><summary>Makefile</summary>

```Makefile
APPLICATION = sleepy_node

# add "DEVELHELP=1" to the call to the make command
# to enable safety checking and error reporting
DEVELHELP=1

BOARD ?= samr21-xpro

# This has to be the absolute path to the RIOT base directory:
RIOTBASE ?= $(CURDIR)/../..

# We neet RTC to wake from hibernate
FEATURES_REQUIRED += periph_rtc

# Modules to include:
USEMODULE += shell_commands

# gnrc is a meta module including all required, basic gnrc networking modules
USEMODULE += gnrc
USEMODULE += gnrc_netdev_default

# manually initialize the network interface
USEMODULE += gnrc_netif_init_devs

# pull in the GNRC IP stack
USEMODULE += gnrc_ipv6_default

# enable GNRC notifications
USEMODULE += gnrc_netif_bus

# only print on error
CFLAGS += -DCONFIG_SKIP_BOOT_MSG

include $(RIOTBASE)/Makefile.include

# Set a custom channel if needed
include $(RIOTMAKE)/default-radio-settings.inc.mk
```
</details>

Some trial & error showed that this behavior only occurs if there is a *second* node that is on the whole time.
This node is just idling with the `gnrc_networking` and did receive a valid public address.

**edit** I could also reproduce the behavior with only one sleepy node and a border router.
**edit2** a second node sure helps in provoking the behavior. ~~This also happens if the second node just uses `gnrc_ipv6_default` instead of `gnrc_ipv6_router_default` and no RPL~~ a second `gnrc_ipv6_router_default` is certainly required for this behavior to occur. 

**edit3:** see https://github.com/RIOT-OS/RIOT/issues/14735#issuecomment-674147653 on how to reproduce this on `native` 

#### Expected results

On success, the above program will produce no output.

#### Actual results

after a couple iterations:
```
2020-08-08 17:20:27,121 #  >> border router timeout <<
2020-08-08 17:20:27,128 # Iface  5  HWaddr: 71:A1  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
2020-08-08 17:20:27,129 #           
2020-08-08 17:20:27,133 #           Long HWaddr: 5A:06:59:C9:2C:36:7D:78 
2020-08-08 17:20:27,141 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-08-08 17:20:27,149 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  6LO  
2020-08-08 17:20:27,150 #           IPHC  
2020-08-08 17:20:27,153 #           Source address length: 8
2020-08-08 17:20:27,156 #           Link type: wireless
2020-08-08 17:20:27,162 #           inet6 addr: fe80::5806:59c9:2c36:7d78  scope: link  VAL
2020-08-08 17:20:27,169 #           inet6 addr: 2001:db8::5806:59c9:2c36:7d78  scope: global  TNT[3]
2020-08-08 17:20:27,171 #           inet6 group: ff02::1
```

and if the *second* node was unplugged

```
2020-08-08 17:24:18,174 #  >> border router timeout <<
2020-08-08 17:24:18,181 # Iface  5  HWaddr: 71:A1  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
2020-08-08 17:24:18,182 #           
2020-08-08 17:24:18,186 #           Long HWaddr: 5A:06:59:C9:2C:36:7D:78 
2020-08-08 17:24:18,193 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-08-08 17:24:18,200 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  6LO  
2020-08-08 17:24:18,202 #           IPHC  
2020-08-08 17:24:18,205 #           Source address length: 8
2020-08-08 17:24:18,207 #           Link type: wireless
2020-08-08 17:24:18,213 #           inet6 addr: fe80::5806:59c9:2c36:7d78  scope: link  VAL
2020-08-08 17:24:18,216 #           inet6 group: ff02::1
```

On the border router:

```
ping ff02::1
12 bytes from fe80::1075:7bff:fe4d:6a7f%5: icmp_seq=0 ttl=64 time=27.100 ms
12 bytes from fe80::1075:7bff:fe4d:6a7f%5: icmp_seq=1 ttl=64 time=19.020 ms
12 bytes from fe80::5806:59c9:2c36:7d78%6: icmp_seq=2 ttl=64 rssi=-51 dBm time=8.265 ms

--- ff02::1 PING statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 8.265/18.128/27.100 ms

fe80::1 dev #5 lladdr 12:75:7B:4D:6A:7F  STALE GC
fe80::1075:7bff:fe4d:6a7f dev #5 lladdr 12:75:7B:4D:6A:7F  REACHABLE GC

Iface  6  HWaddr: F6:FB  Channel: 26  Page: 0  NID: 0x23 
          Long HWaddr: 00:12:4B:00:18:F1:D6:7A 
           TX-Power: 0dBm  State: IDLE 
          AUTOACK  L2-PDU:102  MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::212:4b00:18f1:d67a  scope: link  VAL
          inet6 addr: 2001:db8::212:4b00:18f1:d67a  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:fff1:d67a
          
Iface  5  HWaddr: D2:70:31:23:59:F5 
          L2-PDU:1500  MTU:1500  HL:64  RTR  
          Source address length: 6
          Link type: wired
          inet6 addr: fe80::d070:31ff:fe23:59f5  scope: link  VAL
          inet6 addr: fe80::2  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff23:59f5
          inet6 group: ff02::1:ff00:2
```

#### Versions

Riot Master

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description
On `master` if a node is unreachable once it gets stuck in a loop:
`_handle_snd_ns()` will call `_probe_nbr()` which schedules an `GNRC_IPV6_NIB_SND_MC_NS` event.

This event is handled by `gnrc_ipv6_nib_handle_timer_event()` which in turn calls `_handle_snd_ns()`, starting the whole process all over again.

The result is that only periodic `GNRC_IPV6_NIB_SND_MC_NS` events are generated without ever sending a neighbor solicitation.

The result is that the previously unreachable node is stuck in the TNT[3] state forever.

As a quick fix, just send out the neighbor solicitation instead of scheduling the event.

### Testing procedure

Follow the process outlined in #14735

Basically, have a node that is offline for a while and then tries to re-join the border router.

The [sleepy_node](https://github.com/benpicco/sleepy_node) application will do just that. There is no need to run the CoAP server for this test. Just run a `gnrc_border_router` and this app.

You will find that on `master` after a few cycles, the node will always get a "border router timeout".
With this patch you should instead always get a "get time failed", or if you run the CoAP server as described, the current time.

### Issues/PRs references

This fixes #14735

### CODE REVIEW

**Comment by miri64** [comment]
In https://github.com/RIOT-OS/RIOT/issues/14735#issuecomment-686459959 you report that `next_ns` is suddenly increasing to 4294967295. Shouldn't we rather investigate why that is instead of potentially breaking the algorithm laid out in [RFC 7048](https://tools.ietf.org/html/rfc7048)

**Comment by miri64** [comment]
Let's first clarify your analysis in https://github.com/RIOT-OS/RIOT/issues/14735. See https://github.com/RIOT-OS/RIOT/issues/14735#issuecomment-688113542.

**Comment by benpicco** [comment]
This patch also works

```patch
--- a/sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c
+++ b/sys/net/gnrc/network_layer/ipv6/nib/_nib-arsm.c
@@ -299,7 +299,7 @@ void _probe_nbr(_nib_onl_entry_t *nbr, bool reset)
                     }
                     DEBUG("(retrans. timer = %ums)\n", (unsigned)retrans_time);
                     ipv6_addr_set_solicited_nodes(&sol_nodes, &nbr->ipv6);
-                    _snd_ns(&nbr->ipv6, netif, NULL, &sol_nodes);
+                    _snd_uc_ns(nbr, true);
                     _evtimer_add(nbr, GNRC_IPV6_NIB_SND_MC_NS, &nbr->nud_timeout,
                                  retrans_time);
                     if (nbr->ns_sent < (NDP_MAX_NS_NUMOF + 2)) {
```

**Comment by miri64** [comment]
But that is breaking RFC 7048... Unicasting to a unreachable node is meaningless.

**Comment by benpicco** [comment]
Well *something* is obviously broken here :wink: 
Does a [packet dump](http://benpicco.de/foobar/sleepy_node.pcapng) help?

The node is

```
2020-09-10 17:51:45,418 # Iface  7  HWaddr: 48:C5  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
2020-09-10 17:51:45,419 #           
2020-09-10 17:51:45,427 #           Long HWaddr: 00:04:25:19:18:01:C8:C5 
2020-09-10 17:51:45,431 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2020-09-10 17:51:45,438 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  6LO  
2020-09-10 17:51:45,439 #           IPHC  
2020-09-10 17:51:45,442 #           Source address length: 8
2020-09-10 17:51:45,446 #           Link type: wireless
2020-09-10 17:51:45,451 #           inet6 addr: fe80::204:2519:1801:c8c5  scope: link  VAL
2020-09-10 17:51:45,458 #           inet6 addr: 2001:db8::204:2519:1801:c8c5  scope: global  TNT[3]
2020-09-10 17:51:45,460 #           inet6 group: ff02::1
```

the border router is

```
Iface  6 
          MTU:65535  HL:64  RTR  
          
          Link type: wired
          inet6 addr: fe80::2  scope: link  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff00:2
          
Iface  5  HWaddr: 50:5E  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK 
          
          Long HWaddr: 8F:1B:51:02:E4:02:D0:5E 
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
          ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  RTR  
          RTR_ADV  6LO  IPHC  
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::8d1b:5102:e402:d05e  scope: link  VAL
          inet6 addr: 2001:db8::8d1b:5102:e402:d05e  scope: global  VAL
          inet6 group: ff02::2
          inet6 group: ff02::1
          inet6 group: ff02::1:ff02:d05e
```

**Comment by miri64** [comment]
Wait, so the border router is multicasting, not the node (which is not in the multicast group)?

**Comment by miri64** [comment]
If so:

1. Why is the node set to unreachable in the first place? Registered nodes should only fall into Neighbor Unreachable Detection (and thus become unreachable), when the address registration lifetime times out.
2. Why is the NS sent by the node for the link-local address as target, but with the global address as source? It should be both global.

**Comment by miri64** [comment]
I think we can agree, that this is more of a hack than an actual fix. Please re-open if you disagree.

**Comment by miri64** [comment]
(let's continue discussing in #14735)
